### PR TITLE
Task 3.3

### PR DESCRIPTION
**Pull request by evTyurin**
Complete task 3 after lecture №3

### CODE REVIEW

File: src/com/senlainc/warsaw/tyurin/Turret.java [hunk 1]
```diff
@@ -0,0 +1,9 @@
+package com.senlainc.warsaw.tyurin;
+
+public class Turret implements ILineStep, IProductPart{
```
Comment (mrseneca) [inline src/com/senlainc/warsaw/tyurin/Turret.java]: It's bad design. Object should not be implemented so different interfaces. It's like Turret at the same time assembly line step and part of the product, which this line should create

File: src/com/senlainc/warsaw/tyurin/Engine.java [hunk 2]
```diff
@@ -0,0 +1,9 @@
+package com.senlainc.warsaw.tyurin;
+
+public class Engine implements ILineStep, IProductPart{
```
Comment (mrseneca) [inline src/com/senlainc/warsaw/tyurin/Engine.java]: Same as Turret

File: src/com/senlainc/warsaw/tyurin/Body.java [hunk 3]
```diff
@@ -0,0 +1,9 @@
+package com.senlainc.warsaw.tyurin;
+
+public class Body implements ILineStep, IProductPart{
```
Comment (mrseneca) [inline src/com/senlainc/warsaw/tyurin/Body.java]: Same as Turret and Engine

File: src/com/senlainc/warsaw/tyurin/Tank.java [hunk 4]
```diff
@@ -0,0 +1,30 @@
+package com.senlainc.warsaw.tyurin;
+
+public class Tank implements IProduct, IAssemblyLine{
```
Comment (mrseneca) [inline src/com/senlainc/warsaw/tyurin/Tank.java]: Same as parts but now product, you already should understand what here incorrect

**Review by mrseneca — COMMENTED** [review]
ProductPart, LineStem and Product should have independent implementations like in real live - product cannot build himself and be a part of conveyor and so on

File: src/com/senlainc/warsaw/tyurin/TankProduct.java [hunk 5]
```diff
@@ -0,0 +1,22 @@
+package com.senlainc.warsaw.tyurin;
+
+public class TankProduct implements IProduct{
+    private Body body;
```
Comment (mrseneca) [inline src/com/senlainc/warsaw/tyurin/TankProduct.java]: If you change it to abstraction (interface) class-cast will not be needed lower in code. better always use abstraction

File: src/com/senlainc/warsaw/tyurin/Test.java [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+public class Test {
+
+    public static void main(String[] args) {
+
+        BodyLineStep bodyLineStep = new BodyLineStep();
+        Body body = (Body) bodyLineStep.buildProductPart();
+        System.out.println("Created part of tank - body");
+
+        EngineLineStep engineLineStep = new EngineLineStep();
+        Engine engine = (Engine) engineLineStep.buildProductPart();
```
Comment (mrseneca) [inline src/com/senlainc/warsaw/tyurin/Test.java]: Same here, your StepLines produces abstract IProductPart, if you change variable type here to it class cast will not be needed

File: src/com/senlainc/warsaw/tyurin/TankAssemblyLine.java [hunk 7]
```diff
@@ -0,0 +1,21 @@
+package com.senlainc.warsaw.tyurin;
+
+public class TankAssemblyLine implements IAssemblyLine{
+    private Body body;
+    private Engine engine;
```
Comment (mrseneca) [inline src/com/senlainc/warsaw/tyurin/TankAssemblyLine.java]: You can also use abstraction here

**Review by mrseneca — COMMENTED** [review]
Awesome, just little fix to make your code better

**Review by mrseneca — COMMENTED** [review]
Good job

**Review by mrseneca — APPROVED** [review]

**Comment by evTyurin** [comment]
> Good job

Thank you!

**Pull request closed without merging**